### PR TITLE
Change your own email via the profile view

### LINKED ISSUE #2021: Change your own email via the profile view
## Expected Behaviour

The email input on the profile page should be editable.

## Actual Behaviour

The email input only displays your current email address and can not be changed.

## Steps to Reproduce
1. Access the profile view by selecting your email at the top right of the dashboard.
1. Try to change the email displayed.

## Versions
  - Authoring Tool Version: 0.5.0
  - Framework Version: N/A
  - Operating System: Linux Ubuntu 14.04
  - Browser: Google Chrome

### PR DESCRIPTION
**Pull request by canstudios-nicolaw**
Fixes #2021

### CODE REVIEW

**Comment by canstudios-nicolaw** [comment]
I am closing https://github.com/adaptlearning/adapt_authoring/pull/2022 and replacing it with this. The solution was still the same, it was just easier to make a fresh branch

**Comment by lc-thomasberger** [comment]
Looks good. Can you please add email validation as we have for creating a user.

**Comment by canstudios-nicolaw** [comment]
@lc-thomasberger email validation appears to already be happening as part of the model's validation. Strangely, I can see the validate function in `frontend/src/modules/user/models/userProfileModel.js` but this seems to only check for the presence of an email.

invalidemail

File: frontend/src/modules/user/views/userProfileView.js [hunk 1]
```diff
@@ -105,12 +105,14 @@ define(function(require){
...
     saveUser: function() {
       var self = this;
+      var prevEmail = self.model.get('email');
 
       this.$('.error-text').addClass('display-none');
 
       var toChange = {
         'firstName': self.$('#firstName').val().trim(),
-        'lastName': self.$('#lastName').val().trim()
+        'lastName': self.$('#lastName').val().trim(),
```
Comment (tomgreenfield) [inline frontend/src/modules/user/views/userProfileView.js]: Suggest ditching the quotes on these keys.

File: lib/usermanager.js [hunk 2]
```diff
@@ -293,14 +293,28 @@ exports = module.exports = {
...
+        if (update.email) {
+          // verify the email does not already exist
+          db.retrieve('user', { email: update.email }, function (error, results) {
+            if (error) {
+              return callback(error);
+            }
+            if (results && results.length) {
+              return callback(new DuplicateUserError());
+            } else {
+              db.update('user', search, update, callback);
```
Comment (tomgreenfield) [inline lib/usermanager.js]: No need for `else` when using `return`.

**Comment by dancgray** [comment]
@canstudios-nicolaw, the duplicate username check does not allow the user to change first name, last name or password without also changing the email.

**Comment by canstudios-nicolaw** [comment]
Good spot thanks @dancgray I'll get this updated today

File: lib/usermanager.js [hunk 3]
```diff
@@ -293,14 +293,27 @@ exports = module.exports = {
...
-        }, configuration.getConfig('dbName'));
-      } else {
-        callback(new Error('No matching user record found'));
+        return callback(error);
       }
+      if (!result) {
+        return callback(new Error('No matching user record found'));
+      }
+      database.getDatabase(function(err, db) {
+        if (update.email_prev && update.email !== update.email_prev) {
```
Comment (tomgreenfield) [inline lib/usermanager.js]: `update.email_prev` never seems to be defined for me and so the duplicate user error is not caught:

error

**Comment by taylortom** [comment]
**Found a very minor issue**: the validation messages don't seem to be removed once they're shown. 

In below screenshot, I did the following: 
1. Entered an invalid email, no first name/surname. 
2. On clicking save, I was shown validation errors as expected.
3. After correcting the email, and trying to save again, I was still shown the invalid email label (this applies to any combination of the fields).

Regardless of which messages are shown, you are able to save as expected, so it's not affecting the actual functionality.

screenshot 2019-02-19 at 13 10 54

**Comment by canstudios-nicolaw** [comment]
@taylortom Validation issue should be fixed now :+1:

**Pull request merged**